framework/filters/PageCache.php

```diff
@@ -213,43 +237,71 @@ protected function restoreResponse($response, $data)
      */
     public function cacheResponse()
     {
-        $this->view->endCache();
-        $response = Yii::$app->getResponse();
+        array_pop($this->view->cacheStack);
+        $beforeResult = $this->beforeCacheResponse();
+        if ($beforeResult === false) {
+            $content = ob_get_clean();
+            if (empty($this->view->cacheStack) && !empty($this->dynamicPlaceholders)) {
+                $content = $this->updateDynamicContent($content, $this->dynamicPlaceholders);
+            }
+            echo $content;
+            return;
+        }
+
+        $response = Yii::$app->response;
```

Comment:
The entire `cacheResponse` method does not fit on one screen.
Should be split for a smaller ones.